pkg/controller/controller.go

```diff
@@ -61,32 +62,35 @@ func (prc *PlacementRequestController) ScheduleOne(ctx context.Context, pr *v1al
 		return nil
 	}
 
-	// if there is no binding present then we cannot process the
-	// PlacementRequest. XXX this should be handled at the api level.
-	if len(pr.Spec.Bindings) == 0 {
-		prc.logger.V(3).Info("no bindings found", "obj", prid)
-		pr.Status.Result = v1alpha1.PlacementRequestResultRejected
-		pr.Status.Message = "The request was rejected because it has no bindings"
-		updater := prc.client.KombinerV1alpha1().PlacementRequests(pr.Namespace)
-		_, err := updater.Update(ctx, pr, metav1.UpdateOptions{})
-		return err
-	}
+	// here we create a few shortcuts to api access entities we are going
+	// to use during this function. these shortcuts are already namespace
+	// scoped.
+	prqclient := prc.client.KombinerV1alpha1().PlacementRequests(pr.Namespace)
+	podlister := prc.podlister.Pods(pr.Namespace)
 
-	// we only operate on Lenient policy. for the all or nothing policy a
-	// change in the api server will be necessary.
-	if pr.Spec.Policy != v1alpha1.PlacementRequestPolicyLenient {
-		prc.logger.V(3).Info("unsupported policy", "policy", pr.Spec.Policy, "obj", prid)
+	if err := helpers.Validate(pr); err != nil {
+		prc.logger.Error(err, "placement request is not valid", "obj", prid)
 		pr.Status.Result = v1alpha1.PlacementRequestResultRejected
-		pr.Status.Message = fmt.Sprintf("Unsupported policy: %s", pr.Spec.Policy)
-		updater := prc.client.KombinerV1alpha1().PlacementRequests(pr.Namespace)
-		_, err := updater.Update(ctx, pr, metav1.UpdateOptions{})
+		pr.Status.Message = err.Error()
+		_, err := prqclient.UpdateStatus(ctx, pr, metav1.UpdateOptions{})
 		return err
 	}
 
 	var success bool
 	for _, binding := range pr.Spec.Bindings {
 		prc.logger.V(3).Info("binding pod to node", "bind", binding, "obj", prid)
 
+		if pod, err := podlister.Get(binding.PodName); err != nil {
+			prc.logger.Error(err, "failed to get pod")
+			message := fmt.Sprintf("Failed to get pod %s: %v", binding.PodName, err)
+			helpers.SetPodBindingFailure(pr, binding, "API error", message)
+			continue
+		} else if pod.Spec.NodeName != "" {
+			message := fmt.Sprintf("Pod %s is already bound", binding.PodName)
+			helpers.SetPodBindingFailure(pr, binding, "Already bound", message)
```

Comment:
Should this be a failure if the pod is bound to the intended node?

Reply:
I don't think we should consider it an error when a pod is already bound to a node where it's expected to be. I.e. when the controller gets accidentally rebooted in the middle of the binding of a list of pods.